### PR TITLE
[SPARK-18777][PYTHON][SQL] Return UDF from udf.register

### PR DESCRIPTION
**Pull request by zero323**
## What changes were proposed in this pull request?

- Move udf wrapping code from `functions.udf` to `functions.UserDefinedFunction`.
- Return wrapped udf from `catalog.registerFunction` and dependent methods.
- Update docstrings in `catalog.registerFunction` and `SQLContext.registerFunction`.
- Unit tests.

## How was this patch tested?

- Existing unit tests and docstests.
- Additional tests covering new feature.

### CODE REVIEW

**Comment by holdenk** [comment]
Thank so for doing this, will let me cleanup some of my PySpark notebooks :)

**Comment by holdenk** [comment]
LGTM, since this is just exposing existing functionality I think it makes sense to merge into 2.2 as well, but I'll leave this for a day incase anyone has any comments about this.

**Comment by gatorsmile** [comment]
Python interface `udf.register` needs to be consistent with the scala one.

**Comment by holdenk** [comment]
I don't think the Scala API is particularly consistent and I'd argue that this does actually match the current API as much as is reasonable. Looking at the udf.register interface defined in UDFRegistration.scala this does return either a `UserDefinedAggregateFunction` or `UserDefinedFunction` when passed in UserDefinedAggregateFunction or a Function0 through Function22. depending on which function is passed in most of the time. We do return Unit for the private `registerJavaFunction` & when registering a UDF1 through UDF22 . (Also @gatorsmile this doesn't change the scala API, its really just plumbing through the existing Scala API to Python).

**Comment by gatorsmile** [comment]
In Scala API, we have five different ways to register UDFs. We have to be more careful when changing the external APIs. It makes the users confused if we make the APIs more diverse. I do not suggest to make any further change in Python UDF, until we figure out how to make us easy to add more UDF types.

**Comment by holdenk** [comment]
I agree we need to be careful when making external API changes, but this is a pretty clear win for Python users and it matches the Scala API so I think improving this for Python users shouldn't necessarily block on any unification of the Scala API (is their a plan to improve this or JIRA associated with the planned Scala API changes I could subscribe to notifications on)?

**Comment by gatorsmile** [comment]
So far, one of the biggest issues is we always assume UDF is deterministic. It could cause incorrect results after query optimization. I am working on a UDF API change to add more types for UDFs. Will submit a PR recently. 

We can revisit this PR later.

**Comment by holdenk** [comment]
I feel like that's an unrelated challenge. I'm happy to see other improvements but I'm worried that we will hold up changes for things which aren't happening soon - is there a JIRA for these changes?

**Comment by zero323** [comment]
@gatorsmile This sounds reasonable but I am not sure if I fully understand your concerns.

If anything this brings PySpark closer to the Scala API. At this moment we have

```
 registerFunction(self, name: str, f: Callable[[T], U], returnType: DataType) -> None: ...
```

and we would move to:

```
 registerFunction(self, name: str, f: Callable[[T], U], returnType: DataType) -> Callable[[Column, ...], Column]: ...
```

This, as pointed out by @holdenk, matches `register` API for `Function0` .. `Function22`.

If you're planning breaking changes in the Scala API, it may render this PR obsolete, but we don't commit here to any particular implementation. The only promise here is that registering udf for SQL applications, returns an object, which can be used with `DataFrame` API.  I believe this sounds like a reasonable requirement for any upcoming API.

File: python/pyspark/sql/catalog.py
Comment (HyukjinKwon) [inline python/pyspark/sql/catalog.py]: It looks `strlen` is not being used. Should we remove this or add an example for this?

**Review by HyukjinKwon — APPROVED** [review]
For me, it looks good to me.

**Comment by HyukjinKwon** [comment]
I want to know the UDF API change in more details and see if that blocks this PR. I would appreciate if I can check the details.

File: python/pyspark/sql/catalog.py
Comment (zero323) [inline python/pyspark/sql/catalog.py]: We have to either capture the output (we could use `_` I guess) or use it in the doctests:

```
>>> spark.catalog.registerFunction("stringLengthInt", len, IntegerType())
<function len>
```

I am not sure if it makes sense to use it, unless we provide more diverse tests.

Comment (HyukjinKwon) [inline python/pyspark/sql/catalog.py]: Hm, I think capturing is rather better because it is what this API returns (and as it dose not return `None` anymore) if we don't have a better idea. Variable that indicates it's private, `_`, looks also fine to me.  No strong opinion. I am fine if we don't have a simpler and cleaner idea.

Comment (zero323) [inline python/pyspark/sql/catalog.py]: Let's use `_`. It is a common to indicate discarded output and, unlike `repr`, it is portable.

Comment (HyukjinKwon) [inline python/pyspark/sql/catalog.py]: (Yea, discarded output is a better term ...)

**Comment by gatorsmile** [comment]
@zero323 Will ping you when my PR is submitted. You can then add the UDF changes in Python.

**Comment by gatorsmile** [comment]
@HyukjinKwon @holdenk Spark 2.2 RC has been out. This PR is not a bug fix. Thus, this PR will be merged to the master branch instead of 2.2. I will submit a PR to show the changes soon.

**Comment by gatorsmile** [comment]
JIRA: https://issues.apache.org/jira/browse/SPARK-20586 is opened.

**Comment by zero323** [comment]
Thanks @gatorsmile

**Comment by zero323** [comment]
I think there is no conflict between #17848 and this. As of 2.2 we no longer return `UserDefinedFunction` from `udf` (and we never documented `UserDefinedFunctions`) so changes will have to be reflected in the arguments of `udf` and `UserDefinedFunction.__init__`.

**Comment by HyukjinKwon** [comment]
It sounds orthogonal to me as well. LGTM.

**Comment by HyukjinKwon** [comment]
cc @viirya too who I believe is appropriate to review this.

**Comment by holdenk** [comment]
Thanks for making this PR with the details @gatorsmile it appears to be orthogonal to this change. Historically we've treated Python API parity fixes as closer to bug fixes rather than new features which is why I was thinking this made sense for 2.2, but I'll defer to your expertise in SQL and we can just target this for master.

**Comment by holdenk** [comment]
@gatorsmile want to know if you're ok with this going into master or if you still have concerns about this if its targeted to 2.3?

**Comment by viirya** [comment]
This change LGTM.

I go to check #17848. It seems to me that the PR simply adds two flags into ScalaUDF. It appears that there is not API change regarding with existing UDF registration. I agreed with @holdenk and @HyukjinKwon that it is orthogonal to this change for now.

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Comment by zero323** [comment]
Thanks everyone.

**Pull request closed without merging**
